Readme.txt
SPREAD: A Reliable Multicast and Group Communication Toolkit
-----------------------------------------------------------

/=============================================================================\
| The Spread Group Communication Toolkit - Accelerated Ring Research Version  |
| Copyright (c) 1993-2016 Spread Concepts LLC                                 |
| All rights reserved.                                                        |
|                                                                             |
| The Spread package is licensed under the Spread Open-Source License.        |
| You may only use this software in compliance with the License.              |
| A copy of the license can be found at http://www.spread.org/license         |
|                                                                             |
| This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF       |
| ANY KIND, either express or implied.                                        |
|                                                                             |
| Spread is developed at Spread Concepts LLC with the support of:	      |
|    The Distributed Systems and Networks Lab, Johns Hopkins University       |
|                                                                             |
| Creators:                                                                   |
|    Yair Amir               yairamir@cs.jhu.edu                              |
|    Michal Miskin-Amir      michal@spreadconcepts.com                        |
|    Jonathan Stanton        jonathan@spreadconcepts.com                      |
|    John Schultz            jschultz@spreadconcepts.com		      |
|                                                                             |
| Major Contributors:                                                         |
|    Amy Babay               babay@cs.jhu.edu - accelerated ring protocol     |
|    Ryan Caudy              rcaudy@gmail.com - contribution to process groups|
|    Claudiu Danilov	     claudiu@acm.org - scalable, wide-area support    |
|    Cristina Nita-Rotaru    crisn@cs.purdue.edu - GC security                |
|    Theo Schlossnagle       jesus@omniti.com - Perl, autoconf, old skiplist  |
|    Dan Schoenblum          danschoenblum@gmail.com - Java interface.        |
|                                                                             |
| Contributors:                                                               |
|    Juan Altmayer Pizzorno juan@port25.com - performance and portability     |
|    Jacob Green            jgreen@spreadconcepts.com - Windows support       |
|    Ben Laurie	            ben@algroup.co.uk - FreeBSD port and warning fixes|
|    Daniel Rall            dlr@finemaltcoding.com - Java & networking fixes, |
|                                               configuration improvements    |
|    Marc Zyngier                        - Windows fixes                      |
|                                                                             |
| Special thanks to the following for discussions and ideas:                  |
|    Ken Birman, Danny Dolev, Mike Goodrich, Michael Melliar-Smith,           |
|    Louise Moser, David Shaw, Gene Tsudik, Robbert VanRenesse.               |
|                                                                             |
| Partial funding provided by the Defense Advanced Research Projects Agency   |
| (DARPA) and The National Security Agency (NSA) 2000-2004. The Spread        |
| toolkit is not necessarily endorsed by DARPA or the NSA.                    |
|                                                                             |
| WWW    : http://www.spread.org  and  http://www.spreadconcepts.com          |
| Contact: info@spreadconcepts.com                                            |
|                                                                             |
| Version 5.0.1 built 16/May/2018                                             |
\=============================================================================/

May 16, 2018 Ver 5.0.1
----------------------

Bugfix in membership to drop messages we receive after sending form1 that are not on the form2 token.

February 1, 2017 Ver 5.0.0
--------------------------

No significant changes versus 5.0.0 RC2.

January 23, 2017 Ver 5.0.0 RC2
------------------------------

A new error type, ILLEGAL_TIME, has been exported to the client API and can be returned from SP_connect_timeout if the timeout parameter is invalid.
A new function, E_get_time_monotonic, has been exported to the events API in libspread-util.
Various bugfixes and patches.  See daemon/Changelog for details.

November 22, 2016 Ver 5.0.0 RC1
--------------------------------

IPv6 support fully functional.  See Readme.IPv6 for details.

April 18, 2016 Ver 5.0.0 Alpha
------------------------------

Configuration file parser updated to allow IPv6 addresses, multiple segment addresses and VirtualIDs to be specified.
See Readme.IPv6 for details.

May 27, 2014 Ver 4.4.0
----------------------

No changes versus 4.4.0 RC2.

May 19, 2014 Ver 4.4.0 RC2
--------------------------

Improvements:
  - Slightly improved MSVS project files -- will work with MSBuild 4.0 (e.g. - .NET 4.0, MSVS 2010) now. Later versions of MSVS can auto-upgrade solution (thanks to Johann Lochner)
  - Remove named pipe for accepting UNIX socket connections on exit
  - Allow token sizes up to 64KB (uses IP fragmentation may exacerbate loss -- will log a warning) rather than hard failure if > 1.5KB
  - Send bcast retransmissions immediately rather than queue to suppress unnecessary re-requests in accelerated protocol

Bug fixes:
  - Reset Commit_set to only contain this daemon after installing a regular membership
  - Improved validity check for FORM1 tokens to either process or swallow
  - Only process FORM2 token that corresponds to most recent FORM1 token
  - Block delivery of < AGREED msgs until GROUPS algorithm is completed
  - Consult Last_discarded rather than Aru for deciding whether we have a packet already or not (Prot_handle_bcast)
  - Handle receiving packet after FORM1 processed and it is still listed as a hole on FORM2

January 22, 2014 Ver 4.4.0 RC1
------------------------------
Features:

 - New accelerated ring protocol tailored for data center networks.
   This protocol provides 30%-50% higher throughput and 20-35% lower latency 
   in modern local area networks. Both the original protocol and the accelerated
   ring protocol are available in this version.

 - More efficient packet packing.

 - Windows project files now build again.

 - New spread-service project to make Spread into a proper Windows service.
	- Needs testing from the community!

 - Additional and improved MEMBERSHIP and PROTOCOL logging.

Bug fixes:

 - EVS bugs:
	- Set Aru to 0 when we transition to EVS.
	- Ignore Token->seq when in EVS.
	- Fixed Backoff_membership referring to wrong packet when ring breaks in EVS.
	- Fixed token retransmissions while transitioning from EVS to OP.

 - Token bugs:
	- Token->aru calculaton fixed; Set_aru eliminated.
	- Ignore tokens from wrong membership + sender.

 - Alarm on Windows:
	- Alarm(EXIT) now exits with non-zero code instead of aborting.
	- Fixed long log lines crash bug.

 - Allow ports >= 2^15 to be used.

 - Turned off Nagle algorithm in Java library.

SOURCE INSTALL:
---------------

The source install uses autoconf and make to support many Unix and unix-like 
platforms as well as Windows 95+.

Windows is supported by a set of Visual Studio project files in
the win32 subdirectory. These files build the daemon, libraries, and user
programs.

Any not-specifically supported platform that has reasonably close to normal 
sockets networking should also  be easily portable. See the file PORTING for 
hints on porting.

From the directory where you unpacked the Spread source distribution do 
the following:

1) Run "./configure"

2) Run "make"

3) Run "make install" as a user with rights to write to the selected
installation location. (/usr/local/{bin,man,sbin,include,lib} by default)

4) Create a spread.conf file appropriate to your local configuration. For 
more info on how to do this look at the sample file "sample.spread.conf"
in docs or below in the binary install instructions.

5) To build the java library and documentation, see the Readme.txt file 
in the java directory.

6) To build the perl library see the README file in the perl directory.

BINARY INSTALL:
--------

Packaged binary releases of Spread are available from third parties
for several platforms, such as through the FreeBSD Ports system or as 
RPMs for Linux distributions. The Spread binary release consists of 
prebuilt binaries for several OS's that can be run without being installed
officially into the system.

We recommend that if you are experimenting with spread you create a special
'spread' directory (for example /usr/local/spread or /opt/spread) and keep
all the files together there so it is easy to find stuff.  That also makes 
it easier to run multiple architectures as the binaries for each are in their
own subdirectory. This is not necessary though. You can create that directory
anywhere (e.g. your own directory).

1) Unpack the spread-bin-4.x.tar.gz file into the target directory, 
like /opt/spread

2) Look at the Readme.txt for any updates

3) Look in the bin/{archname} directory for the appropriate architecture for
your system. For example, for Mac OS X 10.3 the binaries to run will be in 
bin/powerpc-apple-darwin7.9.0/

The libraries to link with your application will be found in the
/lib/{archname} directory

The include files to link with are in include/ and the documentation
can be found in docs/

4) To run spread you need to modify the 'docs/sample.spread.conf' file for
   your network configuration.  Comments in the sample file explain this.
   Then cp -p sample.spread.conf /usr/local/bin/spread.conf
   (Notice the name change.  The config file must be named that to be found.
    Alternatively you can run spread -c <config_file_name> )

    You can run "spread usage" to see the spread running options.

To use the Java classes and examples you need to have a copy of the 
main 'spread' daemon running.  Then the class files give you the 
equivelent of the libspread.a as a set of java classes.  The user.java,  
and user.class files give you a demonstration application and source code.  
Under the java/docs directory is a full verison of the Spread Java API
provided by javadocs. 

For Microsoft Windows systems use the spread.exe executable or spread-service.exe
service and the libspread.lib to link with your programs.

OVERVIEW:
---------

Spread consists of two parts, an executable daemon which is executed on 
each machine which is part of the heavyweight 'machine' group, and a library
which provides a programming interface for writing group multicast 
applications and which is linked into the application.

The daemon, called "spread", should be run as a non-priveledged 
user (we created a 'spread' user) and only needs permissions to create a 
socket in /tmp and read its config file which should be in the same 
directory as the executable.  By default the daemon binds to and runs 
off the non-priveledged port 4803 unless the config file indicates otherwise.

Each daemon can be started independently and if it does not find 
any other members of its defined configuration currently active it runs as 
a machine group of 1 machine.  When other daemons are started they will join
this machine group.

The machines which are running a daemon with a common config file 
form a 'machine group' (in contrast to a 'process group').  The daemons 
handle multicasting mesages between each other,providing ordering and 
delivery guarantees, detecting and handling failures of nodes or links, and 
managing all applications which are connected to each daemon.

Each application utilizing the spread system needs to link with 
the 'libspread.a', 'libtspread.a', 'libtspread.lib' or 'libspread.lib' 
library and needs to use the calls defined in the 'sp.h' and 'sp_func.h' 
header files. Documentation for the interface is currenly incomplete, but 
a basic application must do at least the following:

1) Before using any other spread calls you need to 'connect' to a daemon 
by calling the SP_connect(...) call.  This will set a mbox which is an integer 
representing your connection to the daemon.  You use this when making other
spread calls.  You CAN connect to daemons multiple times from the same 
application (and we know of times when this is very useful).

2) If you want to receive messages from a group you need to call SP_join().
Groups are named with standard alphanumeric strings.

3) To send to a group you do NOT need to join it, just call SP_multicast() 
with an appropriate group name.  You will not receive the messages back if 
you are not a member of the group.

4) To be nice to everyone else you should call SP_disconnect() when your 
program is finished using the spread system, if you do not your program's 
disconnection will eventually be detected. 

Some important observations when writing spread programs:

1) All the messages for a particular connection are received during an 
SP_receive() call.  This includes messages from different groups and 
membership messages(like group leave and join), so you must demultiplex 
them yourself.  This is a feature. You can opt to avoid getting membership 
messages all together by indicating that to SP_connect().

2) Spread handles endian differences correctly for metadata and our headers 
but does NOT convert your data for you.  It DOES tell you in the SP_receive 
call whether or not the data originated at machine with a different 
endianness then yours so you can easily use this to convert it yourself 
if necessary. 

3) Each connection to a daemon has a unique 'private group' name which 
can be used to send a message to a particular, and only to a particular, 
application connection.  

4) Spread supports the following ordering/delivery guarantees for messages:

	Unreliable	(least)
	Reliable	(will get there, no ordering)
	Fifo		(reliable and ordered fifo by source)
	Causal		(reliable and all mesg from any source of this level
			are causally ordered)
	Agreed		(reliable and all mesg from any source of this level 
			are totally ordered)
	Safe		(Agreed ordering and mesg will not be delivered to 
			application until the mesg has reached ALL 
			receipients' daemons)

For more detail, much of which IS important, see papers on 
Extended Virtual Sychrony, Transis and Totem.

UTILITY PROGRAMS:
-----------------

	Spread comes with several demonstration programs and useful tools.
They can be found in the main daemon directory as well as the example 
directory. 

1) spmonitor:
	This program has a special interface to spread which allows it to 
control a machine group.  It can terminate all the spread daemons in the 
group, change flow control parameters, and monitor the stats for one or all 
of the machines.  It needs to know the spread.conf file used for a particular
 set of spread daemons.

2) spuser:

	This program is provided with source and uses all the functions of 
the spread interface.  It also acts as a simple client which is useful for 
testing. Reading the source code to this program should show you how to use
all the features of spread, and answer many questions about syntax details.

3) simple_user:

	This is just about the simplest spread program possible and is 
provided with source.  It sends and receives text strings.

4) spflooder:
	
	This is used as one type of performace test.  It 'floods' a spread 
group with data messages. It is provided with source.

5) flush_user:
        This provides a demonstration of using the FL_ interface that provides
Virtual Synchrony semantics to your application. Just like the spuser program
it allows you to join groups and send and receive messages. 

CONTACT:
--------

If you have any questions please feel free to contact:

   spread@spread.org


daemon/Changelog
May 16, 2018 Ver 5.0.1
----------------------

r896 | jschultz | 2018-03-28 01:00:40 -0400 (Wed, 28 Mar 2018) | 5 lines

Attempted bugfix for new packets received after sending form1 that are not reflected on form2
	-- Handle when our Highest_seq is higher than what is on form2
	-- Recompute My_aru due to dropping of such packets (old code didn't do this when we dropped such packets still marked as holes on form2)
Changed some int's -> int32's that are used for indexing packet sequences

February 1, 2017 Ver 5.0.0
--------------------------

r887 | jschultz | 2017-02-01 10:27:15 -0500 (Wed, 01 Feb 2017) | 2 lines
	Updated java's Spread version

January 23, 2017 Ver 5.0.0 RC2
------------------------------

r882 | jschultz | 2017-01-23 07:19:03 -0500 (Mon, 23 Jan 2017) | 5 lines
	Fixed checks against NUM_PRIORITY and NUM_FDTYPES

		Thanks to Daniel Savarese for catching this!

r881 | jschultz | 2017-01-23 07:02:19 -0500 (Mon, 23 Jan 2017) | 5 lines
	daemon/session.c: Delay closing of an operational file descriptor until we are actually removing it, rather than just when we won't use it anymore.
		
		Thanks to Daniel Savarese for this bug report and patch!

r880 | jschultz | 2017-01-23 06:30:41 -0500 (Mon, 23 Jan 2017) | 15 lines
	daemon/Makefile.in: Don't hardcode -ldl in $LIBSPREADUTIL_LIBS, if needed configure should put it in LIBS.
	daemon/arch.h: Addded #include <stdint.h> to pick up INT32_MAX
	daemon/ip_enum.c: Special case FreeBsd to not override scope id from getifaddrs
	daemon/log.c: Reworked logic to only use append mode, not use fseek and ftell, and not add (and overwrite) alive timestamps
	daemon/spread.c: Moved Log_init() to be after any chdir.
	daemon/network.c: Put back special case to not bind to segment address if it is a singleton segment.  Was breaking some old configuration files.
                        Added check to commit suicide if on reconfiguration we are going from singleton to non singleton segment bc on *nix we aren't already bound to segment addr.

	libspread/sp.c: Allow a non-empty send to return 0.  Could cause infinite loop if OS is misbehaving, but other send call was already allowing it.
	libspread-util/buildtools/fixpaths: Removed deprecated defined checks.
	libspread-util/src/data_link.c: Changed IPV6_MULTICAST_HOPS to use an int.  Changed IPV6_MULTICAST_LOOP to use an int, but special case sun to use a char.

		Thanks to Daniel Savarese for these patches!

r879 | jschultz | 2017-01-23 04:08:43 -0500 (Mon, 23 Jan 2017) | 4 lines
	libspread-util/src/data_link.c: Changes to setsockopt parameter types to comply with Solaris / Sun being picky

        	Thanks to Martin Schu for reporting and testing!

r878 | jschultz | 2017-01-23 03:45:40 -0500 (Mon, 23 Jan 2017) | 8 lines
	configure.in: Added checks for poll, poll.h, select, sys/select.h and sin6_len field of sockaddr_in6.  Check for clock_gettime in -lrt.
	config_gram.l: Set sin_len and sin6_len in sockaddr's in parse_ip.
	configuration.c: Suppress an IPv6 specific print when using IPv4.
	sp.h, sess_types.h: Exported a new error type ILLEGAL_TIME.
	sp.c: Rewrote SP_connect* to more cleanly and properly handle signals interrupting I/O.  Can now return ILLEGAL_TIME if user passes an invalid delta time.
	buildtools/fixpaths: Minor change to get rid of deprecated warnings.

r877 | jschultz | 2017-01-23 03:22:11 -0500 (Mon, 23 Jan 2017) | 4 lines
	Exported E_get_time_monotonic
	Minor changes to time functions

November 18, 2016 Ver 5.0.0 RC1 
------------------------------- 

Spread version 5 adds support for IPv6.

Spread version 5 adds support for IPv6 in the Spread daemon, C
language client library and several example and support programs
(e.g. - spmonitor, spuser, spsend, sprecv, spflooder).  This version
of Spread supports Posix environments (e.g. - Linux) and Windows.

Spread 5 can work on IPv4-only hosts, IPv6-only hosts or dual stack
hosts.  However, all daemons in a Spread deployment must run using
only IPv4 or IPv6 addresses.  Spread 5 does not yet support
configurations where some daemons use IPv4 addresses while others use
IPv6 addresses, even if transition mechanisms such as network address
translation (e.g. - stateless IP/ICMP translation) would allow such
hosts to communicate using those addresses.

Daemons' IP addresses can be configured explicitly or by using names
which will be looked up using standard system services (i.e. - DNS,
/etc/hosts, etc.) in order to map them back to IP addresses.  Clients
can similarly specify either explicit addresses or names to connect to
daemons.

r845 | jschultz | 2016-08-08 15:12:02 -0400 (Mon, 08 Aug 2016)
r844 | jschultz | 2016-08-08 12:28:07 -0400 (Mon, 08 Aug 2016)
r843 | jschultz | 2016-08-02 14:48:54 -0400 (Tue, 02 Aug 2016)
r842 | jschultz | 2016-08-02 14:48:08 -0400 (Tue, 02 Aug 2016)
r841 | jschultz | 2016-07-28 12:45:59 -0400 (Thu, 28 Jul 2016)
r840 | jschultz | 2016-07-28 12:37:52 -0400 (Thu, 28 Jul 2016)
r839 | jschultz | 2016-07-18 22:14:26 -0400 (Mon, 18 Jul 2016)
r838 | jschultz | 2016-07-18 22:00:06 -0400 (Mon, 18 Jul 2016)
     Exported a WideAreaNetwork parameter to config file to allow user to optionally specify what set of default membership timeouts they want to use.
     Slight change in setting sin6_scope_id for locally configured ipv6 interfaces
     Added checks that sin6_scope_id's didn't change for local addresses after reconfiguration
     Converted all usages of Alarm to Alarmp
     Downgrade log lines of recving already existing packets from WARNING to DEBUG
     Remove any interface specification from derived name if we derived it from the daemon's address
     Added the ability to specify an interface (scope_id) as part of an IPv6 address using the ::%eth0 syntax.
           - These are excluded from calculating configuration hashes and can therefore be specific to each daemon (as local interface indexes often are) in their configuration file.

daemon/spread_params.h
/*
 * The Spread Toolkit.
 *     
 * The contents of this file are subject to the Spread Open-Source
 * License, Version 1.0 (the ``License''); you may not use
 * this file except in compliance with the License.  You may obtain a
 * copy of the License at:
 *
 * http://www.spread.org/license/
 *
 * or in the file ``license.txt'' found in this distribution.
 *
 * Software distributed under the License is distributed on an AS IS basis, 
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License 
 * for the specific language governing rights and limitations under the 
 * License.
 *
 * The Creators of Spread are:
 *  Yair Amir, Michal Miskin-Amir, Jonathan Stanton, John Schultz.
 *
 *  Copyright (C) 1993-2016 Spread Concepts LLC <info@spreadconcepts.com>
 *
 *  All Rights Reserved.
 *
 * Major Contributor(s):
 * ---------------
 *    Amy Babay            babay@cs.jhu.edu - accelerated ring protocol.
 *    Ryan Caudy           rcaudy@gmail.com - contributions to process groups.
 *    Claudiu Danilov      claudiu@acm.org - scalable wide area support.
 *    Cristina Nita-Rotaru crisn@cs.purdue.edu - group communication security.
 *    Theo Schlossnagle    jesus@omniti.com - Perl, autoconf, old skiplist.
 *    Dan Schoenblum       dansch@cnds.jhu.edu - Java interface.
 *
 */


#ifndef	INC_SPREAD_PARAMS
#define INC_SPREAD_PARAMS

#define		SP_MAJOR_VERSION	5
#define         SP_MINOR_VERSION        0
#define         SP_PATCH_VERSION        1

#define         SPREAD_PROTOCOL         3

#define         SPREAD_BUILD_DATE       "16/May/2018"

#define		DEFAULT_SPREAD_PORT	4803

#ifndef SP_RUNTIME_DIR
#define         SP_RUNTIME_DIR          "/var/run/spread"
#endif
#ifndef SP_GROUP
#define         SP_GROUP                "spread"
#endif
#ifndef SP_USER
#define         SP_USER                 "spread"
#endif
#ifndef SP_UNIX_SOCKET
#define         SP_UNIX_SOCKET          "/tmp"
#endif

#define		MAX_PROC_NAME		 20     /* including the nul terminator, so actually usable is one less, look for it if changed */

#define         MAX_ID_SIZE              16     /* ids are unsigned 32b integer values */

#define		MAX_PROCS_SEGMENT	128
#define		MAX_SEGMENTS		 20
#define		MAX_PROCS_RING		128
#define		MAX_ADDRS_SEGMENT         5
#define         MAX_INTERFACES_PROC      10

#define         MAX_REPS                 25
#define         MAX_FORM_REPS            20

#define		MAX_PACKETS_IN_STRUCT 	8192
#define		PACKET_MASK		0x00001fff

#define		MAX_SEQ_GAP		1600	/* used in flow control to limit difference between highest_seq and aru */

#define		MAX_EVS_ROUNDS		500 	/* used in EVS state to limit total # of rounds to complete EVS */

#define		WATER_MARK		500	/* used to limit incoming user messages */

#define		MAX_PRIVATE_NAME	 10     /* not including the null, look for it if changed */

#define		MAX_GROUP_NAME		 (1+MAX_PRIVATE_NAME+1+MAX_PROC_NAME)
					/* #private_name#proc_name  including the null */
#include        "spu_events.h"
#define		MAX_SESSIONS		( ( MAX_FD_EVENTS-5 ) / 2 ) /* reserves 2 for each connection */

#define		DEFAULT_MAX_SESSION_MESSAGES	1000
#define         MAX_GROUPS_PER_MESSAGE  100     /* Each multicast can't send to more groups then this */

#define         MAX_WRAP_SEQUENCE_VALUE (1<<30) /* Maximum value for token->seq before resetting to zero with membership */

#define		DEFAULT_WINDOW 		   160
#define		DEFAULT_PERSONAL_WINDOW	    20
#define		DEFAULT_ACCELERATED_WINDOW  15

#endif /* INC_SPREAD_PARAMS */
